ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import java.sql.{Connection, PreparedStatement}
+import java.time.Instant
+
+import anorm.{BatchSql, NamedParameter, Row, SimpleSql, SqlStringInterpolation, ToStatement}
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.ledger.EventId
+import com.daml.platform.store.Conversions._
+
+object EventsTablePostgresql extends EventsTable {
+
+  /**
+    * Insertions are represented by a single statement made of nested arrays, one per column, instead of JDBC batches.
+    * This leverages a PostgreSQL-specific feature known as "array unnesting", which has shown to be considerable
+    * faster than using JDBC batches.
+    */
+  final class Batches(
+      insertEvents: Option[SimpleSql[Row]],
+      updateArchives: Option[BatchSql],
+  ) extends EventsTable.Batches {
+    override def execute()(implicit connection: Connection): Unit = {
+      insertEvents.foreach(_.execute())
+      updateArchives.foreach(_.execute())
+    }
+  }
+
+  private val updateArchived =
+    """update participant_events set create_consumed_at={consumed_at} where contract_id={contract_id} and create_argument is not null"""
+
+  private def archive(consumedAt: Offset)(contractId: ContractId): Vector[NamedParameter] =
+    Vector[NamedParameter](
+      "consumed_at" -> consumedAt,
+      "contract_id" -> contractId.coid,
+    )
+
+  override def toExecutables(
+      tx: TransactionIndexing.TransactionInfo,
+      info: TransactionIndexing.EventsInfo,
+      serialized: TransactionIndexing.Serialized,
+  ): EventsTable.Batches = {
+    val capacity = info.events.size
```

Comment:
Perhaps `batchSize`, rather than `capacity`? The latter implies we may never need to reach it. For example, I can have a vector of _capacity_ 1024 but only ever put 100 items in it, and its _size_ will be 100.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b